hello signing samples (#78)
notaryproject/notation: hello-signing.md

# Hello World for Notation

> To get started with Notation, the most basic steps involve:

Author: > ...does the user need to specify the artifact type at this point

The `artifatcType` of the signature isn't specified by the user, rather the notation libraries. 
The user just has to indicate what they're signing, which they only need to do by specifying a `registry/namespace/repo:tag` or digest.

> To get started with Notation, the most basic steps involve:

Reviewer: How does the library detect the artifact type? Or is it the case that notation library is adding artifacts as part of the signing and the artifactType property are for the artifacts we are adding?

1. Acquiring the Notation CLI
2. Generating a Private Key
3. Signing an Artifact
4. Verifying an Artifact with a Notary Signature

This doc outlines progressive scenarios enabling a range of scenarios from the first use to a production deployment.

## Scenarios

To demonstrate how to store and sign a set of supply chain artifacts, we'll walk through a set of scenarios:
- Singing a single container image
- Publishing the container image across public registries, 
- Importing public content to a private registry
- Promoting from dev through production

![](../media/notary-e2e-scenarios.svg)

> - Wabbit Networks is a small software company that produces network monitoring software.

Author: Style choice. Deferring to @nate-double-u for doc input?

> - Wabbit Networks is a small software company that produces network monitoring software.

Reviewer: In either case you might want to add a sentence that introduces the characters in the story. Such as, 

To illustrate these scenarios we will introduce two example companies: _Wabbit Networks_ and _ACME Rockets_.

> - Wabbit Networks is a small software company that produces network monitoring software.

Reviewer 2: I think it's fine to have narrative elements in a list, and agree with @juliusl's introduction suggestion.

- ACME Rockets wishes to acquire network monitoring software.
- ACME Rockets doesn't know about Wabbit Networks, but finds their [net-monitor software in Docker Hub](https://hub.docker.com/r/wabbitnetworks/net-monitor)
- Since they've never heard of Wabbit Networks, they're a little reluctant to run network software without some validations.
- ACME Rockets has a policy to only import Docker Hub certified software, or other approved vendors.
- Wabbit Networks works with Docker Hub to get certified, to help with their customer confidence.
- ACME Rockets will only deploy software that's been scanned and approved by the ACME Rockets security team. They know it's been approved because all approved software has been signed by the ACME Rockets security team.

## Getting Started
- Setup a few environment variables.  
  >Note see [Simulating a Registry DNS Name](#simulating-a-registry-dns-name) to use `registry.wabbit-networks.io`
  ```bash
  export PORT=5000
  export REGISTRY=localhost:${PORT}
  export REPO=${REGISTRY}/net-monitor

>   export IMAGE=${REPO}:v1

Author: Notary signs a specific thing, so we need a tag or a digest. 
As for latest:

> friends don't let friends build against latest

>   export IMAGE=${REPO}:v1

Reviewer: Ah fair enough. Counter to latest: 

> Friends don't push broken bits to latest

  ```
- Install [Docker Desktop](https://www.docker.com/products/docker-desktop) for local docker operations
- Run a local instance of the [CNCF Distribution Registry][cncf-distribution]
  ```bash
  docker run -d -p ${PORT}:5000 ghcr.io/oras-project/registry:latest
  ```
- Acquire the Notation CLI  
Notation releases can be found at: [Notation Releases][notation-releases]  
  ```bash

>   curl -O ~/bin/notation https://github.com/notaryproject/notation/releases/download/v0.0.1-alpha/notation-v0.0.1-alpha.1-linux-amd64

Author: umm, oss here :)

>   curl -O ~/bin/notation https://github.com/notaryproject/notation/releases/download/v0.0.1-alpha/notation-v0.0.1-alpha.1-linux-amd64

Reviewer: bit.ly then? :D

>   curl -O ~/bin/notation https://github.com/notaryproject/notation/releases/download/v0.0.1-alpha/notation-v0.0.1-alpha.1-linux-amd64

Reviewer: _Just to elaborate on my concern/opinion._ 

I have left a lot of comments on broken and outdated links in github documentation. I know it's not up to us to solve the problem. But I think it's in the spirit of OSS if we should ensure working links for getting-started docs like these. 

Aka.ms might be internally allocated, but in this case, I feel it is at the benefit of OSS. An aka.ms link communicates several things:
a) Someone at Microsoft has chosen this link
b) We can find who made this link in case it goes stale
c) We can repair the link without needing to change the repo
d) We're committed to this project by offering these resources, as someone will remain responsible for this link internally

There is precedence for this in OSS. For example, when you install tools to work with GRPC in go: 

```
$ go install **google.golang.org/protobuf/cmd/**protoc-gen-go@v1.26
$ go install **google.golang.org/grpc/cmd/**protoc-gen-go-grpc@v1.1
```

(That being said, it would be nice if we did something similar for `oras`)

>   curl -O ~/bin/notation https://github.com/notaryproject/notation/releases/download/v0.0.1-alpha/notation-v0.0.1-alpha.1-linux-amd64

Author: I completely appreciate the stale issue. If it were an aka.ms link, than only Microsoft folks can edit them. In this case, anyone can make a PR to merge the changes. 
To put this PR into perspective, this is a doc we're using to stabilize the cli and put some scenarios around what/why we're building out the specific CLIs. Nate is building the website through [notaryproject.dev](https://github.com/notaryproject/notaryproject.dev)

I don't expect any of this to be final, and we'll iterate. Particularly the specific releases.
I'm hoping we'll land something more dynamic, or point folks to: https://github.com/notaryproject/notation/releases

  curl -O ~/.docker/cli-plugins/docker-notation https://github.com/notaryproject/notation/releases/download/v0.0.1-alpha/docker-notation-v0.0.1-alpha.1-linux-amd64
  ```

## Building and Pushing
- Build and Push the `net-monitor` software
  ```bash
  docker build -t $IMAGE https://github.com/wabbit-networks/net-monitor.git#main

  docker push $IMAGE
  ```
- List the image, and any associated signatures
  ```bash
  notation list $IMAGE
  ```
  At this point, the results are empty, as there are no existing signatures

## Signing a Container Image

> As you automate the signing of content, you'll most likely want to create and store the private keys in a key vault. Detailed production steps will be covered later. To get things started quickly, the notation cli supports generating self signed key-pairs.

Reviewer: re: "key vault", I think the generic term is key store? Also this paragraph reads better if you flip it. 
```
To get things started quickly, the notation cli supports generating self signed key-pairs. As you automate the signing of content, you'll most likely want to create and store the private keys in a key vault. (Detailed production steps will be covered later)
```
It puts the important info in front of the reader, and if they don't care about production right now they skip over the sentence.

- Generating a Private Key  

>   The following will generate an X-509 private key under the ./notary/ directory

Reviewer: It would be useful to write this line as a comment instead so it's easier to read

  ```bash
# The following will generate an X-509 private key under the ./notary/ directory
  notation cert generate "wabbit-networks.io"
  ```
Also, if someone is copy pasting from here, they get free documentation.

Author: To keep the code to a minimum, I'm working to keep the text outside the copy/paste commands.

  ```bash
  notation cert generate "wabbit-networks.io"
  ```
- Sign the `net-monitor:v1` container image
  ```bash
  notation sign $IMAGE
  ```
- List the image, and any associated signatures
  ```bash
  notation list $IMAGE
  ```
## Verify a Container Image Using Notation Signatures

> To avoid Trojan Horse attacks, it's important to verify the integrity of the artifact, based on an identify of an entity you trust, before pulling the artifact into an environment. Notation uses a set of configured public keys to verify the the content. The `notation cert generate` command automatically added the public key.

Reviewer: This read kind of like a run-on, how about:

To avoid a Trojan Horse attack, before pulling an artifact into an environment, it is important to verify the integrity of the artifact based on the identity of an entity you trust. Notation uses a set of configured public keys to verify the content. The `notation cert generate` command automatically added the public key.

Also, there were two the's in the sentence

> - Attempt to verify the `net-monitor:v1` notation signature

Reviewer: everywhere else you have $IMAGE except here

  ```bash
  notation verify $IMAGE
  ```

>   *The above verification should fail, as you haven't yet configured the keys to trust.*

Reviewer: You should put the caveat as a comment so if someone is following along and copy pasting they can read the story after they've pasted. 

  ```bash
# This line should fail verification, as you haven't yet configured the keys to trust
  notation verify $IMAGE
  ```

Author: Keeping the code to a minimum, so I was keeping all the text in the docs.

- To assure users opt-into the public keys they trust, add the key to the trusted store
  ```bash

>   notation cert add "wabbit-networks.io" ~/.notary/wabbit-networks.io.cert

Reviewer: Ah I missed that line up above, leaving a comment above.

>   notation cert add "wabbit-networks.io" ~/.notary/wabbit-networks.io.cert

Author: > Does .notary get created when you run notation cert generate

Yes, although we'll want to track the question of where things get created.

  ```
- Verify the `net-monitor:v1` notation signature
  ```bash
  notation verify $IMAGE
  ```

>   The above now succeeds as the image is signed with a trusted public key

Reviewer: Same as above and grammar 

  ```bash
 # This should now succeed because the image is signed with a trusted public key
  notation verify $IMAGE
  ```

## Simulating a Registry DNS Name

> Configure the additional steps to simulate a fully qualified dns name for wabbit-networks.

Reviewer: This sentence reads a bit weird, how about, 

Here are the additional steps to simulate a fully qualified dns name for wabbit-networks.

> - Setup names and variables with `registry.wabbit-networks.io`

Reviewer: Setup names and variables _for_ `registry.wabbit-networks.io`

  ```bash
  export PORT=80
  export REGISTRY=registry.wabbit-networks.io
  export REPO=${REGISTRY}/net-monitor
  export IMAGE=${REPO}:v1
  ```

> - Edit the `~/.notary/notation.json` file to support local, insecure registries

Reviewer: Drop the the and file since you've implied this with backticks

Edit `~/.notary/notation.json` to support local insecure registries

# Hello World for Notation

To get started with Notation, the most basic steps involve:

1. Acquire the Notation CLI
2. Generate a Private Key
3. Sign an Artifact

> 4. Verify an Artifact with a Notary Signature

Author: Changed to bullets

- Sign a single container image
- Publish the container image across public registries, 
- Import public content to a private registry
- Promote from dev through production

![](../media/notary-e2e-scenarios.svg)

To illustrate these scenarios we will introduce two example companies:

- Wabbit Networks is a small software company that produces network monitoring software.
- ACME Rockets wishes to acquire network monitoring software.
- ACME Rockets doesn't know about Wabbit Networks, but finds their [net-monitor software in Docker Hub](https://hub.docker.com/r/wabbitnetworks/net-monitor)
- Since they've never heard of Wabbit Networks, they're a little reluctant to run network software without some validations.
- ACME Rockets has a policy to only import Docker Hub certified software, or other approved vendors.
- Wabbit Networks works with Docker Hub to get certified, to help with their customer confidence.
- ACME Rockets will only deploy software that's been scanned and approved by the ACME Rockets security team. They know it's been approved because all approved software has been signed by the ACME Rockets security team.

## Getting Started

> - Setup a few environment variables.  

Reviewer 2: I'd suggest introducing the section before jumping into a list here.

Author: I've updated the flow

  >Note see [Simulating a Registry DNS Name](#simulating-a-registry-dns-name) to use `registry.wabbit-networks.io`
  ```bash
  export PORT=5000
  export REGISTRY=localhost:${PORT}
  export REPO=${REGISTRY}/net-monitor
  export IMAGE=${REPO}:v1
  ```
- Install [Docker Desktop](https://www.docker.com/products/docker-desktop) for local docker operations
- Run a local instance of the [CNCF Distribution Registry][cncf-distribution]
  ```bash
  docker run -d -p ${PORT}:5000 ghcr.io/oras-project/registry:latest
  ```
- Acquire the Notation CLI  
Notation releases can be found at: [Notation Releases][notation-releases]  
  ```bash
  curl -O ~/bin/notation https://github.com/notaryproject/notation/releases/download/v0.0.1-alpha/notation-v0.0.1-alpha.1-linux-amd64
  curl -O ~/.docker/cli-plugins/docker-notation https://github.com/notaryproject/notation/releases/download/v0.0.1-alpha/docker-notation-v0.0.1-alpha.1-linux-amd64
  ```

## Building and Pushing

> - Build and Push the `net-monitor` software

Reviewer 2: Same as above, I'd suggest adding a line or two introducing the section before starting the list here.

  ```bash
  docker build -t $IMAGE https://github.com/wabbit-networks/net-monitor.git#main

  docker push $IMAGE
  ```
- List the image, and any associated signatures
  ```bash
  notation list $IMAGE
  ```
  At this point, the results are empty, as there are no existing signatures

## Signing a Container Image

> To get things started quickly, the notation cli supports generating self signed key-pairs. As you automate the signing of content, you will most likely want to create and store the private keys in a key vault. (Detailed production steps will be covered later)

Reviewer 2: ```suggestion
To get things started quickly, the Notation CLI supports generating self signed key-pairs. As you automate the signing of content, you will most likely want to create and store the private keys in a key vault. (Detailed production steps will be covered later)
```

Reviewer 3: ```suggestion
To get things started quickly, the notation cli supports generating self signed certificates. As you automate the signing of content, you will most likely want to create and store the private keys in a key vault. (Detailed production steps will be covered later)
```

- Generating a Private Key  
  The following will generate an X-509 private key under the `./notary/` directory
  ```bash
  notation cert generate "wabbit-networks.io"
  ```
- Sign the `net-monitor:v1` container image
  ```bash
  notation sign $IMAGE
  ```
- List the image, and any associated signatures
  ```bash
  notation list $IMAGE
  ```
## Verify a Container Image Using Notation Signatures

> To avoid a Trojan Horse attack, before pulling an artifact into an environment, it is important to verify the integrity of the artifact based on the identity of an entity you trust. Notation uses a set of configured public keys to verify the content. The `notation cert generate` command automatically added the public key.

Reviewer 2: ```suggestion
To avoid a Trojan Horse attack, and before pulling an artifact into an environment, it is important to verify the integrity of the artifact based on the identity of an entity you trust. Notation uses a set of configured public keys to verify the content. The `notation cert generate` command automatically added the public key.
```

The "The `notation cert generate` command automatically added the public key." line feels awkward -- should it be "adds" in this case? Also, I'm not sure where it's added to, or when in this process this addition happens.

Reviewer 3: Adding certificate to as trusted for verification should probably be an explicit step, otherwise any test certificates created by user will be automatically added into the verification policy, and be unintentionally used in production if they copied over the same policy.

Author: See #86

- Attempt to verify the $IMAGE notation signature
  ```bash
  notation verify $IMAGE
  ```
  *The above verification should fail, as you haven't yet configured the keys to trust.*
- To assure users opt-into the public keys they trust, add the key to the trusted store
  ```bash
  notation cert add "wabbit-networks.io" ~/.notary/wabbit-networks.io.cert
  ```
- Verify the `net-monitor:v1` notation signature
  ```bash
  notation verify $IMAGE
  ```
  The above now succeeds as the image is signed with a trusted public key

## Simulating a Registry DNS Name

> Here are the additional steps to simulate a fully qualified dns name for wabbit-networks.

Reviewer 2: Should "dns" be all caps here?

Author: I figured since it's part of a title, yes. Since the doc expert asked, I'm assuming it should be lower case. 
Changed

Reviewer 2: I should have been clearer -- I was referring to DNS in the sentence. I think it should be all caps in both places.

Reviewer 2: (I'd also suggest adding a line break above this title)

- Setup names and variables for `registry.wabbit-networks.io`
  ```bash
  export PORT=80
  export REGISTRY=registry.wabbit-networks.io
  export REPO=${REGISTRY}/net-monitor
  export IMAGE=${REPO}:v1
  ```

> - Edit `~/.notary/notation.json` to support local, insecure registries

Reviewer 3: ```suggestion
- Edit `~/.notation/notation.json` to support local, insecure registries
```
Also, we can update the directory structure and file names once we finalize on those.

  ```json
  {
    "enabled": true,
    "verificationCerts": [
    ],
    "insecureRegistries": [
      "registry.wabbit-networks.io"

>     ]

Reviewer 3: What is the intent of specifying insecureRegistries ? Signed artifacts from any source can be verified as long as the signers public key is trusted by the client.

Author: The notation cli uses oras-go for push/pull/verify operations. When working with registries, the default experiences requires https.
The user must opt-into supporting non https registries. 
See #85 as this isn't yet complete.

  ```bash
  #LINUX, including WSL
  curl -Lo notation.tar.gz https://github.com/shizhMSFT/notation/releases/download/v0.5.2/notation_0.5.2_linux_amd64.tar.gz
  tar xvzf notation.tar.gz -C ~/bin notation
  ```

## Building and Pushing
- Build and Push the `net-monitor` software
  ```bash
  docker build -t $IMAGE https://github.com/wabbit-networks/net-monitor.git#main

  docker push $IMAGE
  ```
- List the image, and any associated signatures
  ```bash
  notation list $IMAGE
  ```
  At this point, the results are empty, as there are no existing signatures

## Signing a Container Image

To get things started quickly, the Notation cli supports generating self signed certificates. As you automate the signing of content, you will most likely want to create and store the private keys in a key vault. (Detailed production steps will be covered later)

- Generate a self-signed test certificate for signing artifacts
  The following will generate a self-signed X.509 certificate under the `~/config/notation/` directory
  ```bash
  notation cert generate-test --default "wabbit-networks.io"
  ```
- Sign the container image
  ```bash

>   notation sign $IMAGE

Reviewer 3: I'd prefer being explicit about the key being used and not using a default, we can address that later.

Author: We did want to enable the simple/default scenarios but appreciate the concern. 
@gokarnm, can you open an issue for this?

Reviewer 3: Opened #91

  ```
- List the image, and any associated signatures
  ```bash
  notation list $IMAGE 
  ```

## Verify a Container Image Using Notation Signatures

> To avoid a Trojan Horse attack, and before pulling an artifact into an environment, it is important to verify the integrity of the artifact based on the identity of an entity you trust. Notation uses a set of configured public keys to verify the content. The `notation cert generate-test` command created the public key, however it must be implicitly added for verification to succeed.

Reviewer 3: ```suggestion
To avoid a Trojan Horse attack, and before pulling an artifact into an environment, it is important to verify that the artifact was unmodified after it was created (integrity), and from an trusted entity (authenticity). Notation uses a set of configured public keys that represent trusted entities, to verify the content. The `notation cert generate-test` command created the public key, however it must be implicitly added for verification to succeed.
```

- Attempt to verify the $IMAGE notation signature
  ```bash
  #notation verify $IMAGE
  notation cert remove wabbit-networks.io # see https://github.com/notaryproject/notation/issues/86
  notation verify --plain-http $IMAGE
  ```
  *The above verification should fail, as you haven't yet configured the keys to trust.*
- To assure users opt-into the public keys they trust, add the key to the trusted store
  ```bash
  notation cert add --name "wabbit-networks.io" ~/.config/notation/certificate/wabbit-networks.io.crt
  ```
- Verify the `net-monitor:v1` notation signature
  ```bash
  # notation verify $IMAGE
  notation verify --plain-http $IMAGE
  ```
  This should now succeed because the image is signed with a trusted public key

## Reset
To resetting the environment

- Remove keys, certificates and notation `config.json`  
  `rm -r ~/.config/notation/`
- Restart the local registry  
  `docker rm -f $(docker ps -q)`

> ## Simulating a Registry dns Name
> 
> Here are the additional steps to simulate a fully qualified dns name for wabbit-networks.

Reviewer 2: ```suggestion

## Simulating a Registry DNS Name

Here are the additional steps to simulate a fully qualified DNS name for wabbit-networks.
```
